In [55]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.gridspec as gs
from sklearn.linear_model import LogisticRegression
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

%matplotlib inline
%load_ext autoreload

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [56]:

data = pd.read_csv("C:/Users/ngala/Downloads/Churn_Modelling.csv")
df = data.copy()

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [77]:
df.head()

RowNumber   Surname  CreditScore  Gender       Age  Tenure   Balance  \
0          1  Hargrave        0.538  Female  0.324324     0.2  0.000000   
1          2      Hill        0.516  Female  0.310811     0.1  0.334031   
2          3      Onio        0.304  Female  0.324324     0.8  0.636357   
3          4      Boni        0.698  Female  0.283784     0.1  0.000000   
4          5  Mitchell        1.000  Female  0.337838     0.2  0.500246   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0       0.000000          1               1         0.506735       1  
1       0.000000          0               1         0.562709       0  
2       0.666667          1               0         0.569654       1  
3       0.333333          0               0         0.469120       0  
4       0.000000          1               1         0.395400       0

Above shows the first five rows

In [59]:
df.tail()

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0

Above shows the last five rows

In [80]:

df.drop("Surname",axis="columns",inplace=True)

Dropping unnecessary data

In [81]:
df.head()

CreditScore  Gender       Age  Tenure   Balance  NumOfProducts  HasCrCard  \
0        0.538  Female  0.324324     0.2  0.000000       0.000000          1   
1        0.516  Female  0.310811     0.1  0.334031       0.000000          0   
2        0.304  Female  0.324324     0.8  0.636357       0.666667          1   
3        0.698  Female  0.283784     0.1  0.000000       0.333333          0   
4        1.000  Female  0.337838     0.2  0.500246       0.000000          1   

   IsActiveMember  EstimatedSalary  Exited  
0               1         0.506735       1  
1               1         0.562709       0  
2               0         0.569654       1  
3               0         0.469120       0  
4               1         0.395400       0

In [62]:
df.dtypes

RowNumber            int64
CustomerId           int64
Surname             object
CreditScore          int64
Gender              object
Age                  int64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
dtype: object

In [63]:
pd.to_numeric(df.Balance)

0            0.00
1        83807.86
2       159660.80
3            0.00
4       125510.82
          ...    
9995         0.00
9996     57369.61
9997         0.00
9998     75075.31
9999    130142.79
Name: Balance, Length: 10000, dtype: float64

In [64]:
for column in df:
    print(df[column].unique())

[    1     2     3 ...  9998  9999 10000]
[15634602 15647311 15619304 ... 15584532 15682355 15628319]
['Hargrave' 'Hill' 'Onio' ... 'Kashiwagi' 'Aldridge' 'Burbidge']
[619 608 502 699 850 645 822 376 501 684 528 497 476 549 635 616 653 587
 726 732 636 510 669 846 577 756 571 574 411 591 533 553 520 722 475 490
 804 582 472 465 556 834 660 776 829 637 550 698 585 788 655 601 656 725
 511 614 742 687 555 603 751 581 735 661 675 738 813 657 604 519 664 678
 757 416 665 777 543 506 493 652 750 729 646 647 808 524 769 730 515 773
 814 710 413 623 670 622 785 605 479 685 538 562 721 628 668 828 674 625
 432 770 758 795 686 789 589 461 584 579 663 682 793 691 485 650 754 535
 716 539 706 586 631 717 800 683 704 615 667 484 480 578 512 606 597 778
 514 525 715 580 807 521 759 516 711 618 643 671 689 620 676 572 695 592
 567 694 547 594 673 610 767 763 712 703 662 659 523 772 545 634 739 771
 681 544 696 766 727 693 557 531 498 651 791 733 811 707 714 782 775 799
 602 744 588 747 583 627 731 6

Showing the unique values in the datasets

In [82]:
 
df["Gender"].replace({"Female":1,"Male":0},inplace=True)

Replacing the values in gender to machine readable

In [83]:
df["Gender"].unique()

array([1, 0], dtype=int64)

In [84]:
df.head()

CreditScore  Gender       Age  Tenure   Balance  NumOfProducts  HasCrCard  \
0        0.538       1  0.324324     0.2  0.000000       0.000000          1   
1        0.516       1  0.310811     0.1  0.334031       0.000000          0   
2        0.304       1  0.324324     0.8  0.636357       0.666667          1   
3        0.698       1  0.283784     0.1  0.000000       0.333333          0   
4        1.000       1  0.337838     0.2  0.500246       0.000000          1   

   IsActiveMember  EstimatedSalary  Exited  
0               1         0.506735       1  
1               1         0.562709       0  
2               0         0.569654       1  
3               0         0.469120       0  
4               1         0.395400       0

In [85]:
#scaling the data
cols_to_scale = ["CreditScore","Age","Tenure","Balance","NumOfProducts","EstimatedSalary"]
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

df[cols_to_scale] = scaler.fit_transform(df[cols_to_scale])

In [86]:
df.head()

CreditScore  Gender       Age  Tenure   Balance  NumOfProducts  HasCrCard  \
0        0.538       1  0.324324     0.2  0.000000       0.000000          1   
1        0.516       1  0.310811     0.1  0.334031       0.000000          0   
2        0.304       1  0.324324     0.8  0.636357       0.666667          1   
3        0.698       1  0.283784     0.1  0.000000       0.333333          0   
4        1.000       1  0.337838     0.2  0.500246       0.000000          1   

   IsActiveMember  EstimatedSalary  Exited  
0               1         0.506735       1  
1               1         0.562709       0  
2               0         0.569654       1  
3               0         0.469120       0  
4               1         0.395400       0

In [87]:
x = df.drop("Exited",axis="columns")
y = df["Exited"]

In [88]:
# split the data into training and testing sets
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=5)

In [89]:
x_train.shape

(8000, 9)

In [90]:
x_test.shape

(2000, 9)

In [91]:
# Instantiate a Logistic Regression model without an intercept. C is set to an arbitrarily large number. Use 'liblinear' solver method.
logreg = LogisticRegression(fit_intercept = False, C = 1e12, solver = 'liblinear')
# Fit the model to our X and y training sets
logreg.fit(x_train, y_train)

LogisticRegression(C=1000000000000.0, fit_intercept=False, solver='liblinear')

In [92]:
# Generate model prediction data for train and test sets
y_hat_train = logreg.predict(x_train)
y_hat_test = logreg.predict(x_test)

times the classifir was correct on the training set

In [93]:
# Find residual differences between train data and predicted train data
residuals = np.abs(y_train - y_hat_train)
# Print value counts of our predicted values 
print(pd.Series(residuals).value_counts())
print('----------------------------------')
# Print normalized value counts of our predicted values
print(pd.Series(residuals).value_counts(normalize = True))

0    6455
1    1545
Name: Exited, dtype: int64
----------------------------------
0    0.806875
1    0.193125
Name: Exited, dtype: float64


Results above show that

. 6455Correct(1545incorrect)

. With 80.7% Accuracy

times the classifir was correct on the testing set

In [94]:
# Repeat previous step with test data
residuals = np.abs(y_test - y_hat_test)
print(pd.Series(residuals).value_counts())
print('---------------------------------')
print(pd.Series(residuals).value_counts(normalize = True))

0    1608
1     392
Name: Exited, dtype: int64
---------------------------------
0    0.804
1    0.196
Name: Exited, dtype: float64


Results above show that

. 1608 Correct(392 incorrect)

. With 80.4% Accuracy